controllers/agentpool_controller_tokens.go

```diff
@@ -251,39 +154,62 @@ func (r *AgentPoolReconciler) createSecret(ctx context.Context, ap *agentPoolIns
 		return err
 	}
 
-	ap.log.Info("Reconcile Agent Tokens", "msg", fmt.Sprintf("Kubernetes Secret %q exists", s.Name))
 	return nil
 }
 
 func (r *AgentPoolReconciler) reconcileAgentTokens(ctx context.Context, ap *agentPoolInstance) error {
 	ap.log.Info("Reconcile Agent Tokens", "msg", "new reconciliation event")
 
-	err := r.createSecret(ctx, ap)
-	if err != nil {
+	if err := r.createSecret(ctx, ap); err != nil {
 		ap.log.Error(err, "Reconcile Agent Tokens", "msg", fmt.Sprintf("failed to create a new Kubernetes Secret %s", agentPoolOutputObjectName(ap.instance.Name)))
 		return err
 	}
 
-	removeTokens, err := getTokensToRemove(ctx, ap)
+	agentTokens, err := ap.getTokens(ctx)
 	if err != nil {
 		return err
 	}
-	if len(removeTokens) > 0 {
-		ap.log.Info("Reconcile Agent Tokens", "msg", fmt.Sprintf("removing %d agent tokens from the agent pool", len(removeTokens)))
-		err := r.removeAgentPoolTokens(ctx, ap, removeTokens)
-		if err != nil {
-			return err
+
+	statusTokens := make(map[string]string, len(ap.instance.Status.AgentTokens))
+	for _, t := range ap.instance.Status.AgentTokens {
+		statusTokens[t.Name] = t.ID
+	}
+
+	for _, token := range ap.instance.Spec.AgentTokens {
+		if tokenID, ok := statusTokens[token.Name]; ok {
+			delete(statusTokens, token.Name)
+			if _, ok := agentTokens[tokenID]; ok {
+				delete(agentTokens, tokenID)
+			} else {
+				if err := r.removeToken(ctx, ap, tokenID); err != nil {
+					return err
+				}
+				if err := r.createToken(ctx, ap, token.Name); err != nil {
+					return err
+				}
+			}
+		} else {
+			if err := r.createToken(ctx, ap, token.Name); err != nil {
+				return err
+			}
```

Comment:
I noticed this code is run whether the conditional above is true or not. Maybe we could formulate it like this:

```golang
for _, token := range ap.instance.Spec.AgentTokens {
	if tokenID, ok := statusTokens[token.Name]; ok {
		delete(statusTokens, token.Name)

		if _, ok := agentTokens[tokenID]; ok {
				delete(agentTokens, tokenID)
				continue
		} 

		if err := r.removeToken(ctx, ap, tokenID); err != nil {
			return err
		}
	}

	if err := r.createToken(ctx, ap, token.Name); err != nil {
		return err
	}
}
```
```
```

Reply:
Thanks!